File: scala/nucleotide-count/src/main/scala/DNA.scala
class DNA(dna: String) {
  val valid = Set('A', 'T', 'C', 'G')
  val baseCounts = Map('A' -> 0, 'T' -> 0, 'C' -> 0, 'G' -> 0)

  def count(nucleotide: Char): Either[String, Int] =
    for {
      n <- validNucleotide(nucleotide)
      counts <- nucleotideCounts
    } yield counts(n)

  def nucleotideCounts: Either[String, Map[Char, Int]] =
    invalid match {
      case None => Right(baseCounts ++ dna.groupBy(c => c).mapValues(_.length))
      case Some(nucleotide) => Left(invalidMsg(nucleotide))
    }

  private def validNucleotide(nucleotide: Char): Either[String, Char] =
    Either.cond(valid.contains(nucleotide),
      nucleotide,
      invalidMsg(nucleotide))

  private def invalid: Option[Char] = dna.find(!valid.contains(_))
  private def invalidMsg(nucleotide: Char): String =
   s"invalid nucleotide '$nucleotide'"
}

File: haskell/nucleotide-count/src/DNA.hs
module DNA (nucleotideCounts) where

import Data.Map (Map, fromList, update)
import Data.Maybe (maybe)

nucleotideCounts :: String -> Either String (Map Char Int)
nucleotideCounts xs = doCount baseCounts nucleotides
  where
    baseCounts = fromList [('A', 0), ('T', 0), ('C', 0), ('G', 0)]
    nucleotides = map newNucleotide xs

doCount :: Map Char Int -> [Either String Char] -> Either String (Map Char Int)
doCount counts [] = Right counts
doCount counts (Left s:_) = Left s
doCount counts (Right n:ns) = doCount newCounts ns
  where
    newCounts = update (Just . (+ 1)) n counts

newNucleotide :: Char -> Either String Char
newNucleotide 'A' = Right 'A'
newNucleotide 'T' = Right 'T'
newNucleotide 'C' = Right 'C'
newNucleotide 'G' = Right 'G'
newNucleotide _   = Left "Invalid nucleotide"

